enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,321 @@
+authors:
+  - "@dtantsur"
+reviewers:
+  - "@flaper87"
+approvers:
+  - "@hardys"
+  - "@zaneb"
+creation-date: 2022-03-14
+last-updated: 2022-03-14
+tracking-link:
+  - https://issues.redhat.com/browse/METAL-10
+status: implementable
+---
+
+# Converge ZTP and Metal3 flows
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes converging two somewhat different bare-metal
+provisioning flows in OpenShift:
+
+- *Standard* Metal3 flow used in bare-metal IPI and [baremetal-operator][bmo].
+  Uses the [Ironic agent][ipa] (also known as ironic-python-agent or
+  simply IPA) for writing the CoreOS image.
+
+- ZTP (Zero-Touch Provisioning) flow using the so-called *live ISO* Ironic and
+  baremetal-operator feature, the assisted service and the assisted agent.
+
+## Motivation
+
+Having two different provisioning flows comes with significant maintenance and
+development costs:
+
+- Some features, namely firmware and RAID configuration, require using the
+  Ironic agent, which is not present in the *live ISO* flow.
+
+- Supporting (i)PXE boot with the *live ISO* flow requires creating new API
+  in baremetal-operator specifically for this purpose.
+
+- In the *live ISO* flow, it is not possible to notify Ironic (and thus
+  baremetal-operator) when the actual installation finishes. From the
+  perspective of Metal3 components, the installation is finished once the ISO
+  is connected. As a result, Ironic cannot disconnect the ISO from the
+  machine's BMC, nor can it update the BMC's boot device settings properly.
+
+- The *live ISO* flow disables Ironic *inspection* process and has to simulate
+  it by updating the `BareMetalHost` based on the assisted service's inspection
+  process. This creates an unnessesary coupling between the assisted service
+  and the baremetal-operator's hardware inventory format (which, in turn, is
+  derived from the Ironic's inventory format).
+
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+- Any different path through the `BareMetalHost` state machine makes it
+  exponentially more difficult for upstream developers to reason about the
+  effects of any changes. In addition, the *live iso* flow is not tested
+  upstream by the Metal3 project.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+
+### Non-Goals
```

Comment:
A non-goal is supporting per-host network configuration through the BMH PreprovisioningNetworkData API.

Reply:
Should probably also document "Perform firmware config changes only _after_ running host validations" as a non-goal, since if it were a goal then this plan wouldn't be sufficient.